pkg/api/resource/quantity.go

```diff
@@ -0,0 +1,406 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"math/big"
+	"regexp"
+	"strings"
+
+	"speter.net/go/exp/math/dec/inf"
+)
+
+// Quantity is a fixed-point representation of a number.
+// It provides convenient marshaling/unmarshaling in JSON and YAML,
+// in addition to String() and Int64() accessors.
+//
+// The serialization format is:
+//
+// <serialized>      ::= <sign><numeric> | <numeric>
+// <numeric>         ::= <digits><suffix> | <digits>.<digits><suffix>
+//   (Note that <suffix> may be ""!)
+// <sign>            ::= "+" | "-"
+// <digits>          ::= <digit> | <digit><digits>
+// <signedDigits>    ::= <digits> | <sign><digits>
+// <digit>           ::= 0 | 1 | ... | 9
+// <suffix>          ::= <binarySuffix> | <decimalExponent> | <decimalSuffix>
+// <binarySuffix>    ::= i | Ki | Mi | Gi | Ti | Pi | Ei
+// <decimalSuffix>   ::= m | "" | k | M | G | T | P | E
+// <decimalExponent> ::= "e" <signedDigits> | "E" <signedDigits>
+//   (Note that 1024 = 1Ki but 1000 = 1k; I didn't choose the capitalization.)
+//
+// No matter which of the three exponent forms is used, no quantity may represent
+// a number greater than 2^63-1 in magnitude, nor may it have more than 3 digits
+// of precision. Numbers larger or more precise will be capped or rounded.
+// (E.g.: 0.1m will rounded up to 1m.)
```

Comment:
SI suffix for thousand is lowercase k, but for 1024 is Ki. As the comment says a few lines up, I didn't make up this system.

As for rounding; it can be changed. I made the current choice under the assumption that you don't want your numbers rounded down to zero. It is easy to change this to do "ordinary" rounding if that's what people find more natural.